### PR TITLE
chore: Add RFC for health endpoint improvements

### LINKED ISSUE #9460: Healthcheck API RFC
RFC for #4250

### PR DESCRIPTION
**Pull request by spencergilbert**
Signed-off-by: Spencer Gilbert <spencer.gilbert@datadoghq.com>

Closes #9460

[Rendered](https://github.com/vectordotdev/vector/blob/spencer/healthcheck-rfc/rfcs/2021-12-01-9460-health-endpoint-improvements.md)

### CODE REVIEW

**Comment by spencergilbert** [comment]
I think a big _tl;dr_ here is it's very easy for us to push events out of Vector's durability setup by marking the instance "unhealthy", which is something we need to be very aware of as we consider the "healthiness" of Vector.

**Review by binarylogic — CHANGES_REQUESTED** [review]
I started reviewing this and backed out to discuss changing the context of this RFC to K8s readiness probes, which appear to correctly cover all aspects "improving Vector's reliability in Production" as it relates to the `health` endpoint. This RFC feels too laser focused on the `health` endpoint without addressing the broader concerns that probes cover.

I don't think this changes the RFC too much. Adjusting the perspective to be from probes would offer the following benefits:

1. Readiness probe -> improves start up and shut down behavior
2. Liveliness probe -> improves self-healing
3. Startup probe -> this doesn't seem applicable to us, right?

It would be nice to see:

1. The probe configuration for each
2. How the `health` endpoint is used and how it should behave, informing its improvements 

Does that make sense?

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+It does open the door for further health check improvements which are more
+complex. Any further improvements to the check need to be planned and discussed
+to avoid unwanted complexity.
+
+## Prior Art
+
+- [Beats](https://www.elastic.co/guide/en/beats/filebeat/7.15/http-endpoint.html)
+  - [Liveness and Readiness Probes](https://github.com/elastic/helm-charts/blob/main/filebeat/values.yaml#L174-L198)
+- [FluentBit](https://docs.fluentbit.io/manual/administration/monitoring#rest-api-interface)
+  - [Liveness and Readiness Probes](https://github.com/fluent/helm-charts/blob/main/charts/fluent-bit/values.yaml#L131-L139)
```
Comment (jszwedko) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements:65]: Very small nit, it's best to link to specific SHAs to ensure the code is always what you meant to link to. The `fluent-bit` one has already been updated in a way that it is one line off.

**Review by jszwedko — COMMENTED** [review]
Great start, thanks for getting this conversation going.

I'll defer further review until Ben's feedback has been addressed, which I think makes sense and also won't be that drastic of an update to the RFC.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements (hunk 1 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: Resolved in ef8a799

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+## Proposal
+
+### User Experience
+
+The existing health endpoint on our GraphQL based API is enhanced to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: Which behavior in particular, 503 specifically or _not_ responding 200 at all times?

[the remark this answers is not in the document]
Comment (binarylogic) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: Yes, 503s specifically, or how K8s responds to different codes.

[the remark this answers is not in the document]
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]:
```
Any code greater than or equal to 200 and less than 400 indicates success. Any other code indicates failure.
```

https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-a-liveness-http-request

`503` seemed most appropriate from established status codes.

**Comment by spencergilbert** [comment]
For what it's worth I deliberately wrote the RFC to not be 100% Kubernetes specific - I think we should establish ahead of time the "platform scope" when doing these. 

> 1. Readiness probe -> improves start up and shut down behavior
> 2. Liveliness probe -> improves self-healing
> 3. Startup probe -> this doesn't seem applicable to us, right?
> 
> It would be nice to see:
> 
> 1. The probe configuration for each
> 2. How the `health` endpoint is used, how it should behave, and any changes/improvements we need to make

Realistically I think Vector should report itself as "unhealthy" as little as possible. We're designing the application in such a way that it should be one of the most (if not the most) reliable/durable sections of your observability pipeline. If we mark a Vector Aggregator as unready via a Readiness probe we stop receiving traffic from agents and push the durability requirements back to the agents at the edge, rather than letting Vector buffer/short-circuit to object storage/double write to an available sink/etc.

1. Readiness - we should report unready if Vector is unable to receive events, regardless of sink status
2. Liveness - I don't really see this as necessary as Vector tends to exit itself if it gets into a non-recoverable state, if we don't think this is the case we should either improve Vector's behavior (in the style of Erlang, let it crash - gracefully handling acks and buffered events) or look to expose a signal of "I'm making no progress on receiving or sending events"
3. Startup - `vector validate` suits this probe, more so than readiness IMHO - as a sink being unhealthy shouldnt (again IMHO) necessarily mark the instance as unusable

**Comment by binarylogic** [comment]
> For what it's worth I deliberately wrote the RFC to not be 100% Kubernetes specific - I think we should establish ahead of time the "platform scope" when doing these.

It's not about the platform, it's about looking at the `health` endpoint from a wholistic perspective of reliability. K8s is just a good lens to look at it through since they have a well defined spec with their probes.

In other words, we don't care about the `health` endpoint. We care about proper load-balancing, startup, shutdown, and self-healing behavior.

**Comment by spencergilbert** [comment]
> It's not about the platform, it's about looking at the `health` endpoint from a wholistic perspective of reliability. K8s is just a good lens to look at it through since they have a well defined spec with their probes.
> 
> In other words, we don't care about the `health` endpoint. We care about proper load-balancing, startup, shutdown, and self-healing behavior.

OK, I think I see what you're getting at. My opinion on the RFC doesn't change regardless, and the usage and downsides of probes remain what I listed above. I'm happy to move those into the RFC but if we're going to discuss them the PR comments seem suitable to me.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements (hunk 2 above)
[the remark this answers is not in the document]
Comment (jszwedko) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: Just noting also that the `/health` route is actually outside of the GraphQL API which is mounted at `/graphql`. This is an important note because I think it means we can use traditional REST semantics here and use status codes to indicate health rather than being restricted by GraphQL which would prefer all requests return 200 and include the health information in the body of the response.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+- Today's health endpoint causes 502's when running behind a loadbalancer
+- The healthcheck isn't integrated with any components and only represents if Vector
+itself is running
+
+## Proposal
+
+### User Experience
+
+The existing health endpoint on our GraphQL based API is enhanced to return 503's
```
Comment (jszwedko) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:32]:
```suggestion
The existing health endpoint on Vector's API is enhanced to return 503's
```

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+these for the global health endpoint. Looking at a single Vector instance from
+a high level this is similar to calculating a simple up/down metric for a
+distributed system. We would need to determine how many errors/failures per
+component cause them to be "unhealthy", and depending on the component and config
+this could be very situational. This would require user provided configuration
+and increase operational complexity, something we intentionally avoid when possible.
+
+## Outstanding Questions
+
+- Should we assign a unique error code for "shutting down"?
```
Comment (jszwedko) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:89]: This seems like something a GraphQL-based health endpoint could support. As noted in my [other comment](https://github.com/vectordotdev/vector/pull/10297#discussion_r766173780) the `/health` endpoint is outside of the GraphQL API. A GraphQL health request might look like a request to `/graphql` with:

```graphql
query Health {
  status
}
```

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+- Do we want to return "unhealthy" if some number of components are "unhealthy"
+  - Depending on deployment and configuration this could push durability concerns
+to an agent layer, which wouldn't necessarily be ideal
+  - This could be introduced down the line as a configurable option
+- Related to the previous, should sinks (or other components) regularly rerun
+their configured healthchecks or should it continue to be at startup only
```
Comment (jszwedko) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:95]: This seems like a separate concern unless we wire up that concept of health to `/health`.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+- Using `vector validate` as a Kubernetes Readiness Probe could take an instance
+out of service discovery and push buffering and durability to a less optimal layer
+  - This is probably a better startup check but may not be ideal for a regular
+runtime check, if that's the case we should use a Startup Probe
```
Comment (jszwedko) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:99]: Looking at https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/ for startup probes, I don't think `vector validate` is a good fit since it exiting 0 doesn't actually indicate that Vector is started. It seems like the startup probe should just be the same check as the readiness probe.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+this could be very situational. This would require user provided configuration
+and increase operational complexity, something we intentionally avoid when possible.
+
+## Outstanding Questions
+
+- Should we assign a unique error code for "shutting down"?
+- Do we want to return "unhealthy" if some number of components are "unhealthy"
+  - Depending on deployment and configuration this could push durability concerns
+to an agent layer, which wouldn't necessarily be ideal
+  - This could be introduced down the line as a configurable option
```
Comment (jszwedko) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:93]: This seems related to the alternative you identified above. I would say we can defer this for now.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 6 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:99]: I was more thinking we'd exit if the sinks/sources weren't available - but that's a good point. Perhaps a reason for `vector health` 🤷

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 5 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:95]: Agreed, I feel like this is worth doing regardless - but isn't directly related here otherwise.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements (hunk 2 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: Ah, I think I misunderstood some of the organization of the code here. 👍

**Review by jszwedko — APPROVED** [review]
👍 I think this is a fine step forward. Having it return unhealthy when shutting down seems pretty uncontroversial.

When we originally scheduled this I think we were thinking that we'd do something more comprehensive like having components report their own health and exposing that (see https://github.com/vectordotdev/vector/issues/4250) but I'm ok with deferring that work for now until we have more need.

**Review by silversupreme — APPROVED** [review]
👍🏻  on this. I think it's a simple win, and easy to get better operational behavior from Vector in common deployment environments. Some of the framing might be good to update, to @binarylogic's thoughts, but the concrete implementation will undoubtedly remain the same since it hews so close to standards.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements (hunk 2 above)
[the remark this answers is not in the document]
Comment (leebenson) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: To give a little background on why `/health` is currently coupled with the API:

The code in src/api spins up a warp HTTP server. I added /health originally as a quick route to check for 'aliveness', and followed that up with a `{ health }` equivalent in GraphQL. Both were kept so that a user could choose which is more relevant (the former potentially being simpler to parse, and not requiring any GraphQL knowledge.)

As `/health` evolves to distinguish between aliveness and readiness, I think it ultimately makes sense to do the same in GraphQL. Instead of returning a boolean, we could return an enum with `Ready`, `NotReady`, etc.

That's probably out of scope for this first go, but just something to keep in mind. It'd be useful to reuse the same mechanism for determining health in the GraphQL API.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+## Scope
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+
+### Out of scope
+
+- Moving the health endpoint out of our existing API
```
Comment (leebenson) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:19]: I'd go further than deeming this out out of scope, and suggest this should necessarily remain coupled for valid technical reasons.

As I [mentioned here](https://github.com/vectordotdev/vector/pull/10297#discussion_r769391071), this was bundled with API code originally because both share a common Warp HTTP server. If we were to move it out of API, that would suggest we'd either need _two_ HTTP servers (complicating the port setup) - or, some other HTTP abstraction that would allow modules such as API and health to separately register routes. I think the latter would be an unnecessary complication.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+### Implementation
+
+We have existing logic in place for other components to handle Vector's shutdown.
+The [health handler](https://github.com/vectordotdev/vector/blob/master/src/api/handler.rs#L7)
+can be updated to respond 503 if shutdown has started and 200 otherwise. This
```
Comment (leebenson) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:43]: Is there a cause to consider issuing a 503 during start-up, too?

Despite shutdown generally being the longer phase, there are situations were start-up can also be relatively expensive.

Third-party API key validation, using a `[provider]` block to retrieve a remote configuration, an eventual enterprise remote loading strategy, etc. These may be asynchronous concerns per-component, but they can add up to wall clock delays between a Vector process being started and it being ready to ingest events.

We could move HTTP initialization above config validation and issue a 503 until config is parsed, to get a more accurate readiness indicator.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+should be a minor change and not negatively impact existing deployments.
+
+## Rationale
+
+This improvement is low effort but greatly improves the experience of running
+Vector behind a load balancer or in a service discovery system. As we reccomend
```
Comment (leebenson) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:49]:
```suggestion
Vector behind a load balancer or in a service discovery system. As we recommend
```

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+both of those options for production deployments there's very little reason not
+to implement this.
+
+## Drawbacks
+
+This is a minor change overall and does not add any real engineering burden.
```
Comment (leebenson) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:55]: This may be out of scope for the initial RFC, but it's worth noting there's still _some_ engineering work.

Potentially:

- Moving API init above config validation (if we want 503s on startup). This is a little more involved than shifting a few lines further up the page; config validation occurs within `Application::prepare_from_opts` and the API is separately spawned in `Application::run`, with a dependency on config already being available. This would need adjusting if we want to issue 503s pre-readiness.

- Additional signal handling to determine when Vector is starting up or shutting down. Even if this winds up just being shutdown, we need some way of updating the health state so that it responds with 503. Currently this is a idempotent/immutable route.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+- Related to the previous, should sinks (or other components) regularly rerun
+their configured healthchecks or should it continue to be at startup only
+- Using `vector validate` as a Kubernetes Readiness Probe could take an instance
+out of service discovery and push buffering and durability to a less optimal layer
+  - This is probably a better startup check but may not be ideal for a regular
+runtime check, if that's the case we should use a Startup Probe
+
+## Plan Of Attack
+
+- [ ] Integrate health endpoint with our shutdown sequence, letting the API return
```
Comment (leebenson) [inline rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:103]: It might be worth noting whether the GraphQL API is explicitly in/out of scope, such that this only concerns `/health` for now or whether it also includes `{ health }` (with appropriate types).

Ideally, we wind up with a state where it's simple to integrate with both without requiring separate signal mechanisms.

**Review by leebenson — CHANGES_REQUESTED** [review]
I like the incremental strategy here of a binary 200/503 signalling ready/not ready. This is a good step forward.

Left a couple of comments regarding whether this should be extended to start-up and some more explicitness regarding the plan for keeping the GraphQL API in sync.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 8 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:19]: 👍 I talked some with @jszwedko last night about conditionally attaching the graphql handler so we could always expose the health endpoint but not the graphql paths, which some users have expressed concerns over having enabled for security reasons.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 12 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:103]: I'd advocate removing the graphql endpoint as much as possible here due to the general concerns I've seen from users looking to run Vector "securely" and keeping the "basic" health endpoint as _dumb_ as possible. Is there a reason why we have both `/health` and `{ health }` today?

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 9 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:43]: I'd think adjusting _when_ the warp server starts is better than having it start and return 503's until "Vector is ready" (which, IMHO isn't well defined/queryable today). We'd still be considered "down" for service discovery mechanisms and achieve the same goal.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 11 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:55]: Yeah, certainly some "hand waving" on difficulty and maintainability here. It's definitely less involved than per component "health" and calculating some health score out of that 😄

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 9 above)
Comment (leebenson) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:43]: Currently, warp starts _after_ a configuration has been loaded (and in some cases, retrieved), so if we don't want to distinguish between "Vector has started but is not yet ready to receive/ingest events" and "Vector is ready", that's already the current state.

However, if we don't care about that distinction during start-up, I'd wonder why we care during shutdown. Why not just take `/health` offline entirely to mirror the 'not reachable' state during start-up?

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 12 above)
Comment (leebenson) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:103]: > Is there a reason why we have both /health and { health } today?

Mostly for the [reasons mentioned here](https://github.com/vectordotdev/vector/pull/10297#discussion_r769391071).

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 8 above)
Comment (leebenson) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:19]: Agreed, `api = false` & `health = true` would be useful.

I still think the coupling makes sense. Like `playground = false`, it'd just be a case of turning on/off certain routes.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 9 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:43]: That's essentially what the RFC proposes, no? Today during shutdown we're still reachable and still responding as "up".

I think the API can be valuable during shutdown for monitoring purposes and don't think I'd recommend shutting down the API prior to the rest of the components.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements (hunk 2 above)
[the remark this answers is not in the document]
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: 🤔 I'm personally still not sold on the need for both the HTTP and GraphQL based health checks, it feels duplicative to me - but I'm also less familiar with GraphQL and have been reaching for things more like `vector validate` than the GraphQL endpoint as I've envisioned this moving forward. (That also seems easier from the kubernetes frame of reference compared to making GraphQL queries)

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 9 above)
Comment (leebenson) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:43]: My main point is that I feel that "startup" and "shutdown" are somewhat equivalent.

They both describe a state in which the Vector _process_ is healthy (in the context of a running binary), but ingestion is inactive or winding down toward inactive.

If we are going to return 503 for one of them, it feels like we should return 503 for the other for completeness.

There _is_ a nuance between both states, which I guess could be an argument against it- during startup, a config could fail to be retrieved or parse, where shutdown _generally_ follows a health config that has been explicitly told to stop. I don't think that's always the case, though (a component can fail and therefore cause a shutdown.), so this still feels somewhat equivalent.

I'm not overly opinionated on this point. It just feels incomplete to accommodate shutdown but not startup.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements (hunk 2 above)
[the remark this answers is not in the document]
Comment (leebenson) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements]: The GraphQL API provides general observability into a running Vector instance. An argument for including health data is that it complements the other categories of data it already provides: host and internal metrics, and component meta. It also serves a different class of client.

A GraphQL client that is querying for those things could extend a query to include health/status, too. An example use case could be `vector top`; I could see displaying a "Shutting down..." notice in the terminal. By including it in the API, this could even be subscribed to rather than polled.

`/health`, OTOH, feels more suited to general purpose readiness checks.

TL;DR - Two different clients, with typically different needs.

File: rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md (hunk 12 above)
Comment (jszwedko) [reply rfcs/2021-12-01-9460-healthcheck-endpoint-improvements.md:103]: I'm a fan of having both. The GraphQL endpoint could support, for example, `vector top` reporting Vector health status.

**Comment by spencergilbert** [comment]
@jszwedko @leebenson thanks for the GraphQL focused feedback 😄 I'll give this us an update later today. I'm not sure how necessary it is to 503 during startup, though I do understand the symmetry argument.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md [hunk 13]
```diff
@@ -0,0 +1,117 @@
+# RFC 9460 - 2021-12-01 - Health endpoint improvements
+
+Our existing `/health' endpoint is limited and additional information should be
```
Comment (binarylogic) [inline rfcs/2021-12-01-9460-health-endpoint-improvements.md:3]:
```suggestion
Our existing `/health` endpoint is limited and additional information should be
```

**Review by tobz — APPROVED** [review]
This all looks pretty solid and reasonable to me.  We can always improve the health endpoint further in future RFCs/enhancements, and this RFC alone will make it far more useful to Kubernetes right out of the gate.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+The existing health endpoint on Vector's API is updated to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+The API will not start until the topology is build and validated as functional,
+this isn't precisely when Vector's configured sources _can_ actually process
+events but it's a reasonable first step to ensure graceful startups.
```
Comment (binarylogic) [inline rfcs/2021-12-01-9460-health-endpoint-improvements.md:46]: This makes me think that propagating source's health is the correct solution since both are dependent on sources and sources already implement startup/shutdown logic.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+### `vector health` subcommand
+
+While not suitable for general load balancer usage, it's very easy to `exec`
+commands in Kubernetes to determine health. For prior art we don't have to look
+further than the Datadog Agent which has a subcommand that outputs current
+health information that can be used by operators/systems.
+
+## Outstanding Questions
+
+- Do we need to update the `{ health }` GraphQL query at the same time?
```
Comment (binarylogic) [inline rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: Yes, shouldn't both endpoints use the same code path?

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md (hunk 14 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:46]: Yes, that is part of the future work and specification definition.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md (hunk 15 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: I do not believe they currently are, and regardless I don't think they need to serve the same purpose. If they do serve the same purpose I'd argue that one of them should be removed entirely. cc @leebenson

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+- ~~Should we assign a unique error code for "shutting down"?~~
```
[the remark this answers is not in the document]
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:98]: Response body isn't an error code, I think it's fine to provide more information in the response body (when we have it from components). I don't think it's necessary to create a special error code for shutting down.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md (hunk 15 above)
Comment (leebenson) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: I'm okay with leaving `{ health }` in the API as-is, and tackling later if needed.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+- Improve visibility of high level status at runtime
+
+### Out of scope
+
+- Immediate integrations to expose component level healh
```
Comment (leebenson) [inline rfcs/2021-12-01-9460-health-endpoint-improvements.md:23]:
```suggestion
- Immediate integrations to expose component level health
```

**Review by leebenson — APPROVED** [review]
This seems like a simple and incremental way to introduce health checks 👍🏻 

One thing I'd like to see a little more detail on is how you'd intend to signal shutdown to the [health handler](https://github.com/vectordotdev/vector/blob/master/src/api/handler.rs#L5-L7).

I could see various mechanisms working here, such as static/global state, or perhaps a watch channel. Knowing the intended implementation would allow for more specific feedback, especially in how we might re-use that mechanism when it comes to updating the GraphQL API. It may also inspire a design pattern for communicating individual component status later.

I'm fine with merging as-is and tackling as an implementation step, however.

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md (hunk 15 above)
Comment (binarylogic) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: I'd prefer to remove it if they are going to deviate in behavior. I can't think of a use case for that GraphQL query anyway. Typically this is used for load balancing purposes.

**Review by binarylogic — APPROVED** [review]

File: rfcs/2021-12-01-9460-health-endpoint-improvements.md (hunk 15 above)
Comment (spencergilbert) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: @binarylogic referencing @leebenson's thoughts on this previously: https://github.com/vectordotdev/vector/pull/10297#discussion_r769672177

Comment (leebenson) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: Do we actually make use of the health query in the `top` / `tap` clients:

https://github.com/vectordotdev/vector/blob/7f8e41c026a4b24b81d21984766ffd423b1b38a8/lib/vector-api-client/src/client.rs#L30-L49

Comment (leebenson) [reply rfcs/2021-12-01-9460-health-endpoint-improvements.md:97]: (That last comment should have read "We do", not "Do we". A statement, not a question 😄 )
